### ISSUE TITLE
Components: Validated form components should cancel browser invalid handling

### ISSUE THREAD

**Issue by aduth** [WordPress/gutenberg#73015]
## What problem does this address?

Validated form components use [browser-native form validation](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constraints) behavior to control and intercept form validation to show visible error messages.

However, they don't currently work very well when external application logic programatically invokes validation, such as through [`HTMLFormElement#reportValidity`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/reportValidity). Use-cases include external revisions to a form field's value such as asynchronous validation or mirrored input values.

**Steps to reproduce:**

1. Go to https://wordpress.github.io/gutenberg/?path=/docs/components-validatedtextcontrol--docs
2. Enter "error" into the field
3. Right-click the field and "Inspect element" to inspect the input in your browser DevTools
4. In your DevTools console, enter `$0.reportValidity()`

The above steps result in both the custom visual error label and the browser-native error popup appearing:

<img width="325" height="192" alt="Image" src="https://github.com/user-attachments/assets/c7adac70-28c7-4427-b954-702d859d89fa" />

## What is your proposed solution?

It may be as simple as calling `event.preventDefault()` on [the handled '`invalid'` event](https://github.com/WordPress/gutenberg/blob/03e3101bd7e1a06067c5275ad5a4cf3873603c54/packages/components/src/validated-form-controls/control-with-error.tsx#L109-L112).

Via [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/reportValidity) (emphasis mine):

>The reportValidity() method of the [HTMLFormElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement) interface performs the same validity checking steps as the [checkValidity()](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/checkValidity) method. In addition, for each [invalid](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/invalid_event) event that was fired **and not canceled**, the browser displays the problem to the user.

cc @WordPress/gutenberg-components

**Comment by mirka** [comment]
This is kind of by design. Also it will occur more frequently than "when external application logic programatically invokes validation", because it will be invoked when the user tries to submit the enclosing form.

There was some initial discussion with designers as to whether we were ok with the native popovers. I remember trying to figure out a way to suppress the popover only (but keep the other behaviors like scroll to and focus), but I don't think I found any. It's an interesting trade-off, because the popover actually provides a very convenient and obvious alert to the problem (better than just our custom red message), when the user's attention is not necessarily near the invalid inputs. It's just that we don't have control over the popover's visual design.

Even if we do figure out a way to suppress the native popovers, I think we may need to design and implement some other way to bring attention to the invalid field, which will likely also be a popover given the diversity of UI designs it needs to support.

That's the context of how we decided to go with the native behavior for our first iteration. We can go full custom if we're ready to commit the added effort.

**Comment by aduth** [comment]
Thanks for the context @mirka ! Though it surprises me that we think the messages are an inferior indicator compared to the native popovers. As a user seeing something like in the screenshot, I'd also be confused / distressed about seeing an alert indicator under the popup, but the error text being obscured until I shift focus (_if_ I understand that shifting focus would allow me to read it unobscured).

>We can go full custom if we're ready to commit the added effort.

This would be my soft preference, though I'm not sure what added effort this entails. What about them is insufficient as-is?

**Comment by mirka** [comment]
> What about them is insufficient as-is?

I think the point is that the native popovers are a "last resort" alert mechanism here. The user is basically given softer warning cues _before_ clicking on the submit button, whether that be "required/optional" indicators on the field labels, or the red error messages that have already appeared while filling out the form. In the majority case, the native popovers won't ever be seen if the user has noticed these cues and corrected the validation errors before attempting to submit.

Whether the user simply glossed over the cues, or if there are other ability issues (e.g. color blindness, forced-colors mode, cannot interpret the tiny warning icon), we don't know. But given that the cues already failed to catch their attention on the way down to the submit button, that probably justifies an escalation to a stronger, different form of alert.

> As a user seeing something like in the screenshot, I'd also be confused / distressed about seeing an alert indicator under the popup, but the error text being obscured

This is understandable. Maybe there are ways to increase the prominence of the errors without resorting to an obscuring popover.

> I'm not sure what added effort this entails

Aside from the design effort, at the very least we'll need to add logic for the "focus on first invalid field" functionality. Basically add a `submit` event listener on the enclosing form to find the first invalid control in `form.elements`. This would either need to be done by the control components themselves (where they each add a listener to `elementRef.current.form` that works in an idempotent way), or through a designated `Form` component.

**Comment by aduth** [comment]
Personally, I don't know that I agree with the premise that the current validation treatment (red border, alert icon, red error message) is "soft", or at least that using multiple different ways of informing the user about the error is the best way to handle that. I'd be more inclined to something more like you mention with "increase the prominence of the errors without resorting to an obscuring popover", maybe building on the focus effect for fields with errors if the native form submission will shift focus to the problematic field.

Default|Focused
---|---
<img width="528" height="107" alt="Image" src="https://github.com/user-attachments/assets/c1f04a7f-ead8-4f50-9903-6e5e34164460" />|<img width="523" height="97" alt="Image" src="https://github.com/user-attachments/assets/f66dc643-3e0b-4b12-95cf-b186ad6b38d0" />

**Comment by mirka** [comment]
Part of the complexity I've noticed is that not all control UI designs have an inherent focus ring we can easily repurpose for an error indicator. These can require case-by-case design assessment on whether simply changing the existing focus ring red has sufficient prominence, and that it doesn't "look weird", especially against any theme accent colors used. For example checkbox groups, radio groups, sliders, and probably other custom composed controls that we haven't added validation to yet.

I'm hoping we can land on a universal design that works for any control UI without requiring extra design decisions or implementation per component. I have to give it to the native implementation in this regard — the popovers will just work universally, no interference with existing colors or layout. I think this consideration will become increasingly relevant once consumers start adding validation to their [own custom components](https://github.com/WordPress/gutenberg/pull/72048#discussion_r2467182622).

**Comment by aduth** [comment]
What should the next step be here?

- Pursue this as-is?
- Repurpose this issue (or spin out a prerequisite issue) for one or both of:
   - ... a more generalized solution for increasing emphasis of error messages?
   - ... a consistent, universal UI for input controls that includes error consideration?

**Comment by mirka** [comment]
I guess we can start by hiding the native popovers, and see how it goes? We can always escalate from there, or revert or whatever.

I played around a bit in a sandbox with plain elements, and if things are really as simple as they seem, it might even be as easy as:

```js
// Do this inside the 'invalid' event listener

  e.preventDefault();

  if (!e.target.form) {
    e.target.focus();
    return;
  }
  
  const firstErrorInForm = Array.from(e.target.form.elements).find(el => !el.validity.valid);
  
  if (firstErrorInForm === e.target) {
    e.target.focus();
  }
```

We may find some quirks and edge cases in actual implementation that doesn't allow it to be that simple, but we can assess then whether it's worth the trouble.

**Comment by aduth** [comment]
Ooh, I hadn't even considered that `preventDefault` would also prevent the default _focusing_ behavior that we want to preserve. I recall [handling this in another project I worked on](https://github.com/18F/identity-idp/blob/77a4479c47ce5b768a2b1ce7eb87e78abc7b3a70/app/javascript/packages/validated-field/validated-field-element.ts#L50) by having every element try to focus itself when invalid, but only if the currently-focused element is not associated with an error, with the idea being that every handler after the first would stop. But on reflection there might be some issues with event binding order where it might not guarantee "first" as "earliest in the page".

**Comment by juanfra** [comment]
I'm in favor of leaning on the custom validation. It gives us more control over the visual side and feels more consistent, which ultimately creates a better experience. 

We should just be mindful of the a11y implications as we centralize things. Perhaps we can loop in people from the accessibility team to have a second look at this.
